README.rst
MP 620 - 630 | Debian Based Univeral Installer
##############################################

This should Work on ALL Debian Based Systems

This Installer has been built for x86 and x86\_64. The Installer has been tested on :
  * Ubuntu 8.04 - 12.04
  * Debian 5 - 6 
  * Mint 10 - 12.

  The Universal install has been updated to Version 5.0 and now provides a function to clean up old packages, additionally there is no more compiling, the installer uses all precompiled DEBS which will allow for easy upgrading later as well as better capabillity to uninstall if need be.

  As with my previous post, this article will help you install the Canon MP620 and or MP630 Printer on a Debian Based Linux machine. What differentiates this HOW-TO from the others is that this is a place to download an installer script to allow you to install the printer with minimal effort and by minimal I mean that the Printer in normally functional in around 15 minutes. 

This Installer works on BOTH 32bit and 64bit
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

I developed this script to help ease the installation and transition from Windows to Linux, and to allow people to get their Linux systems running with the least amount of effort.

! Upgrade Process !
-------------------

If you are upgrading to your installation of the Printer drivers / PPD all you now have to do is install the new version. The script will do all of the package purging and PPD removal for you everything else is automatically taken care of.

--------------

Installing the printer
^^^^^^^^^^^^^^^^^^^^^^

The first thing that you need to do is to open a terminal. Go to from the terminal enter the following commands.

.. code-block:: bash

   sudo su -

You will need to go to where you downloaded the repository, once there tell your system that the file is Executable

.. code-block:: bash

    sudo chmod +x install.sh

Now all you need to do is run the installer

.. code-block:: bash

    ./install.sh

If you know your ROOT users password and do not feel like using SUDO, you can simply switch to root for the install of the script. 

.. code-block:: bash

    su -c 'bash ./install.sh'

The installer will take around 5 - 15 minutes to complete. Once Completed you will have to go to your system printer configuration screen.  You should have the canon BJNP protocol available. Click the network discovered Printer, either Canon MP620 or 630 Allow the configuration window to search the for the Drivers The Installer knows what the printer is, if you choose to change the name you can feel free, but the name is auto prompted. Finish the installer and print the test page and close the Printer Configuration Window Upon Competion you will have the ability to print and Scan via USB and or Networking.

--------------

Scanning from the MP620 - 630
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

After Completing the installation your printer will have been setup
and configured for Printing as Well as Scanning.

To user your scanner you have to activate the scan process from the
Linux Host system.

***Your Printer will not see the Linux System from the On-Printer-Menues***

Here is the Simple Scan Process

Open Simple Scan, this can be located in the Graphics Menu, If you
do not have it already you can find Simple Scan in the software
center.
Once Simple Scan is open click Document >> Preferences
Make Sure you see the Canon MP620 Printer Selected, if it is not
selected Select it.

Choose the settings that you want and then click close
Now go to the printer and load a document on the scanning bed
Go back to the Linux System and click the Scan Button
**That is IT!**

--------------

Supplementary Information
-------------------------

Some users have reported that in the 64Bit, x64, installation they
have had to input the IP address of the printer in the "Canon
Networking field" of the Printer install. If your system is not
detecting the Canon Printer in the detected Printers field you can
input the IP address manually

Select Canon Networking type 
  * A bar will open up and allow you to input the Printer protocol, IP and port
  * The input should look like: ``bjnp://X.X.X.X:8611`` (*Replace "X" with the IP address of the Printer*)
  * Now click forward through the menu and make sure you select the Canon Driver set with the "MP630 ver.3.00" Driver. Complete the installation and print a test page.

--------------

Hope this helps
---------------

If this helps you out please let me know and post a comment. I love hearing from you guys.


bjnp/NEWS
2008-08-07 First beta release of CUPS-bjnp.

This release adds a number of additions over the alpha versions released before on fazant.net:
- Adds spec files for Fedora and Centos 5 (RHEL5)
- Compiles without the need for the CUPS sources
- Supports CUPS 1.2 and 1.3 with the same source files

Ink level support is not included but planned to be added to Turboprint

2008-08-09 Second beta release
- added out of paper support
- improved logging/debugging support:
  Normally we log to the cups error log
  For debugging of printer discover we can set env. var BJNP_DEBUG
  For debugging of printing we can set the debuglevel as part of the device URI:
      bjnp://printer-1.pheasant:8611/?debuglevel=DEBUG

  Debugging levels are named as in CUPS

2008-11-18 Fourth beta release
- Added keep-alive for TCP connection as printer may time out if for any reason (slow filter)
  connection is inactive for 30 sec (reportd by ZEDO)
- added a 15 sec delay at end of printjob as printer may otherwise not accept new printjob
  (also reported by Zedo)
- Rewrote specfile so it works for all redhat based releases

2009-02-11 Version 0.5
- Corrected a bug where the backend would crash during printer discovery when it encountered 
  a point to point link.
- Added autoconf/automake support.

2009-05-29 0.5.4
- This release only changes the rpm-spec handling in preparation for release in Fedora.
  There are no other functional changes. So for most people there is no need to update.



bjnp/bjnp.c
/*
 *   bjnp backend for the Common UNIX Printing System (CUPS).
 *   Copyright 2008 by Louis Lagendijk
 *
 *   heavily based on cups AppSocket sources
 *   Copyright 2007 by Apple Inc.
 *   Copyright 1997-2007 by Easy Software Products, all rights reserved.
 *
 *   These coded instructions, statements, and computer programs are the
 *   property of Apple Inc. and are protected by Federal copyright
 *   law.  Distribution and use rights are outlined in the file "LICENSE.txt"
 *   "LICENSE" which should have been included with this file.  If this
 *   file is missing or damaged, see the license at "http://www.cups.org/".
 *
 *   This file is subject to the Apple OS-Developed Software exception.
 *
 * Contents:
 *
 *   main()    - Send a file to the printer or server.
 *   side_cb() - removed and integrated in main loop of RunLoop
 *   wait_bc() - removed as bjnp does not have a true backchannel
 *               it is used to send acks only
 */

/*
 * Include necessary headers.
 */

#include <stdarg.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "bjnp.h"

#ifdef WIN32
#  include <winsock.h>
#else
#  include <unistd.h>
#  include <fcntl.h>
#  include <sys/socket.h>
#  include <netinet/in.h>
#  include <arpa/inet.h>
#  include <netdb.h>
#endif /* WIN32 */


/*
 * 'main()' - Send a file to the printer or server.
 *
 * Usage:
 *
 *    printer-uri job-id user title copies options [file]
 */

int				/* O - Exit status */
main (int argc,			/* I - Number of command-line arguments (6 or 7) */
      char *argv[])		/* I - Command-line arguments */
{
  char method[255],		/* Method in URI */
    hostname[1024],		/* Hostname */
    username[255],		/* Username info (not used) */
    resource[1024],		/* Resource info (not used) */
   *options,			/* Pointer to options */
   *name,			/* Name of option */
   *value,			/* Value of option */
    sep;			/* Option separator */
  int print_fd;			/* Print file */
  int copies;			/* Number of copies to print */
  time_t start_time;		/* Time of first connect */
  int recoverable;		/* Recoverable error shown? */
  int contimeout;		/* Connection timeout */
  int waiteof;			/* Wait for end-of-file? */
  int port;			/* Port number */
  char portname[255];		/* Port name */
  int delay;			/* Delay for retries... */
  int device_fd;		/* AppSocket */
  int error;			/* Error code (if any) */
  int i;			/* loop variable */
  http_addrlist_t *addrlist,	/* Address list */
   *addr;			/* Connected address */
  char addrname[256];		/* Address name */
  ssize_t tbytes;		/* Total number of bytes written */
  char *bjnp_debugstr;		/* environment string */
#if defined(HAVE_SIGACTION) && !defined(HAVE_SIGSET)
  struct sigaction action;	/* Actions for POSIX signals */
#endif /* HAVE_SIGACTION && !HAVE_SIGSET */


  /*
   * Make sure status messages are not buffered...
   */

  setbuf (stderr, NULL);

  /*
   * Ignore SIGPIPE signals...
   */

#ifdef HAVE_SIGSET
  sigset (SIGPIPE, SIG_IGN);
#elif defined(HAVE_SIGACTION)
  memset (&action, 0, sizeof (action));
  action.sa_handler = SIG_IGN;
  sigaction (SIGPIPE, &action, NULL);
#else
  signal (SIGPIPE, SIG_IGN);
#endif /* HAVE_SIGSET */

  /*
   * get debug level for printer discovery based on environment settings
   */

  if ((bjnp_debugstr = getenv ("BJNP_DEBUG")) != NULL)
    bjnp_set_debug_level (bjnp_debugstr);

  /*
   * Check command-line...
   */

  if (argc == 1)
    {
      struct printer_list printers[256];
      int num_printers;
      int i;

      if ((num_printers = bjnp_discover_printers (printers)) == 0)
	puts ("network bjnp \"Unknown\" \"Canon network printer\"");
      else
	for (i = 0; i < num_printers; i++)
	  {
	    printf ("network bjnp://%s:%u \"%s\" \"%s %s\" \"%s\"\n",
		    printers[i].hostname,
		    printers[i].port,
		    printers[i].model,
		    printers[i].model,
		    printers[i].hostname, printers[i].IEEE1284_id);
	  }

      return (CUPS_BACKEND_OK);
    }
  else if (argc < 6 || argc > 7)
    {
      fprintf (stderr, "cups BJNP backend - version %s\n", VERSION);
      _cupsLangPrintf (stderr,
		       _
		       ("Usage: %s job-id user title copies options [file]\n"),
		       argv[0]);
      return (CUPS_BACKEND_FAILED);
    }

  /*
   * If we have 7 arguments, print the file named on the command-line.
   * Otherwise, send stdin instead...
   */

  if (argc == 6)
    {
      print_fd = 0;
      copies = 1;
    }
  else
    {
      /*
       * Try to open the print file...
       */

      if ((print_fd = open (argv[6], O_RDONLY)) < 0)
	{
	  perror ("ERROR: unable to open print file");
	  return (CUPS_BACKEND_FAILED);
	}

      copies = atoi (argv[4]);
    }

  /*
   * Extract the hostname and port number from the URI...
   */

  httpSeparateURI (HTTP_URI_CODING_ALL, cupsBackendDeviceURI (argv),
		   method, sizeof (method), username, sizeof (username),
		   hostname, sizeof (hostname), &port,
		   resource, sizeof (resource));

  if (port == 0)
    port = 8611;		/* Default to bjnp-1 */

  /*
   * Get options, if any...
   */

  waiteof = 1;
  contimeout = 7 * 24 * 60 * 60;

  if ((options = strchr (resource, '?')) != NULL)
    {
      /*
       * Yup, terminate the device name string and move to the first
       * character of the options...
       */

      *options++ = '\0';

      /*
       * Parse options...
       */

      while (*options)
	{
	  /*
	   * Get the name...
	   */

	  name = options;

	  while (*options && *options != '=' && *options != '+'
		 && *options != '&')
	    options++;

	  if ((sep = *options) != '\0')
	    *options++ = '\0';

	  if (sep == '=')
	    {
	      /*
	       * Get the value...
	       */

	      value = options;

	      while (*options && *options != '+' && *options != '&')
		options++;

	      if (*options)
		*options++ = '\0';
	    }
	  else
	    value = (char *) "";

	  /*
	   * Process the option...
	   */

	  if (!strcasecmp (name, "waiteof"))
	    {
	      /*
	       * Set the wait-for-eof value...
	       */

	      waiteof = !value[0] || !strcasecmp (value, "on") ||
		!strcasecmp (value, "yes") || !strcasecmp (value, "true");
	    }
	  else if (!strcasecmp (name, "contimeout"))
	    {
	      /*
	       * Set the connection timeout...
	       */

	      if (atoi (value) > 0)
		contimeout = atoi (value);
	    }
	  else if (!strcasecmp (name, "debuglevel"))
	    {
	      bjnp_set_debug_level (value);
	    }
	}
    }
  
  /* 
   * print command line arguments to debug log
   * We can not do that sooner, as we first need to initialize 
   * the debug system
   */
 
  for (i = 0; i < argc; i++)
    {
      bjnp_debug(LOG_DEBUG, "cups-bjnp: argv[%d] = %s\n", i, argv[i]);
    }

  /*
   * Then try to connect to the remote host...
   */

  recoverable = 0;
  start_time = time (NULL);

  sprintf (portname, "%d", port);

  if ((addrlist = httpAddrGetList (hostname, AF_UNSPEC, portname)) == NULL)
    {
      _cupsLangPrintf (stderr, _("ERROR: Unable to locate printer \'%s\'!\n"),
		       hostname);
      return (CUPS_BACKEND_STOP);
    }

  _cupsLangPrintf (stderr,
		   _("INFO: Attempting to connect to host %s on port %d\n"),
		   hostname, port);

  fputs ("STATE: +connecting-to-device\n", stderr);

  for (delay = 5;;)
    {
      if ((addr = bjnp_send_job_details (addrlist, argv[2], argv[3])) == NULL
	  || (addr = httpAddrConnect (addrlist, &device_fd)) == NULL)
	{
	  error = errno;
	  device_fd = -1;

	  if (getenv ("CLASS") != NULL)
	    {
	      /*
	       * If the CLASS environment variable is set, the job was submitted
	       * to a class and not to a specific queue.  In this case, we want
	       * to abort immediately so that the job can be requeued on the next
	       * available printer in the class.
	       */

	      _cupsLangPuts (stderr,
			     _
			     ("INFO: Unable to contact printer, queuing on next "
			      "printer in class...\n"));

	      /*
	       * Sleep 5 seconds to keep the job from requeuing too rapidly...
	       */

	      sleep (5);

	      return (CUPS_BACKEND_FAILED);
	    }

	  if (error == ECONNREFUSED || error == EHOSTDOWN ||
	      error == EHOSTUNREACH)
	    {
	      if (contimeout && (time (NULL) - start_time) > contimeout)
		{
		  _cupsLangPuts (stderr,
				 _("ERROR: Printer not responding!\n"));
		  return (CUPS_BACKEND_FAILED);
		}

	      recoverable = 1;

	      _cupsLangPrintf (stderr,
			       _
			       ("WARNING: recoverable: Network host \'%s\' is busy; "
				"will retry in %d seconds...\n"), hostname,
			       delay);

	      sleep (delay);

	      if (delay < 30)
		delay += 5;
	    }
	  else
	    {
	      recoverable = 1;

	      _cupsLangPrintf (stderr, "DEBUG: Connection error: %s\n",
			       strerror (errno));
	      _cupsLangPuts (stderr,
			     _
			     ("ERROR: recoverable: Unable to connect to printer; "
			      "will retry in 30 seconds...\n"));
	      sleep (30);
	    }
	}
      else
	break;
    }

  if (recoverable)
    {
      /*
       * If we've shown a recoverable error make sure the printer proxies
       * have a chance to see the recovered message. Not pretty but
       * necessary for now...
       */

      fputs ("INFO: recovered: \n", stderr);
      sleep (5);
    }

  fputs ("STATE: -connecting-to-device\n", stderr);
  _cupsLangPrintf (stderr, _("INFO: Connected to %s...\n"), hostname);

#ifdef AF_INET6
  if (addr->addr.addr.sa_family == AF_INET6)
    fprintf (stderr, "DEBUG: Connected to [%s]:%d (IPv6)...\n",
	     httpAddrString (&addr->addr, addrname, sizeof (addrname)),
	     ntohs (addr->addr.ipv6.sin6_port));
  else
#endif /* AF_INET6 */
  if (addr->addr.addr.sa_family == AF_INET)
    fprintf (stderr, "DEBUG: Connected to %s:%d (IPv4)...\n",
	     httpAddrString (&addr->addr, addrname, sizeof (addrname)),
	     ntohs (addr->addr.ipv4.sin_port));

  /*
   * Print everything...
   */

  tbytes = 0;

  while (copies > 0 && tbytes >= 0)
    {
      copies--;

      if (print_fd != 0)
	{
	  fputs ("PAGE: 1 1\n", stderr);
	  lseek (print_fd, 0, SEEK_SET);
	}

      tbytes = bjnp_backendRunLoop (print_fd, device_fd, addr);

      if (print_fd != 0 && tbytes >= 0)
	{
#ifdef HAVE_LONG_LONG
	  _cupsLangPrintf (stderr,
			   _("INFO: Sent print file, %lld bytes...\n"),
			   CUPS_LLCAST tbytes);
#else
	  _cupsLangPrintf (stderr,
			   _("INFO: Sent print file, %ld bytes...\n"),
			   CUPS_LLCAST tbytes);
#endif /* HAVE_LONG_LONG */
	}
    }


  /*
   * Close the socket connection...
   */

  close (device_fd);

  /*
   * and tell printer to finsh job 
   */
  bjnp_finish_job (addr);

  httpAddrFreeList (addrlist);

  /*
   * delay a bit as otherwise next job may hang (reported by Zedonet for PIXMA MX7600) 
   */ 

  sleep(15);

  /*
   * Close the input file and return...
   */

  if (print_fd != 0)
    close (print_fd);

  if (tbytes >= 0)
    _cupsLangPuts (stderr, _("INFO: Ready to print.\n"));

  return (tbytes < 0 ? CUPS_BACKEND_FAILED : CUPS_BACKEND_OK);
}

/*
 * End of "$Id: socket.c 6911 2007-09-04 20:35:08Z mike $".
 */

